std/traits.d

```diff
@@ -8664,3 +8666,152 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+// Gets the type of a value, or just the type passed if it's a type.
+private template TypeOf(T...)
+{
+    static if (is(T[0] U))
+        alias TypeOf = U;
+    else static if (is(typeof(T[0]) U))
+        alias TypeOf = U;
+}
+
+@safe unittest
+{
+    static assert(is(TypeOf!3 == int));
+    static assert(is(TypeOf!int == int));
+    static assert(is(TypeOf!([]) == void[]));
+    static assert(!is(TypeOf!(mixin(__MODULE__))));
+    static assert(is(TypeOf!TypeOf));
+}
+
+/**
+ * Get a list of the overloads for the given callable, in the order in which they are defined in the code.
+ *
+ * Params:
+ *     func = The callable whose overload set to return.
+ * Returns:
+ *     A tuple of the overloads for the given callable.
+ */
+template getOverloads(func...)
+if (func.length == 1 && isCallable!func)
+{
+    enum isModuleOrPackage(alias s) = !is(typeof(s));
+    alias Parent = __traits(parent, func);
+
+    static if (is(Parent) || isModuleOrPackage!Parent)
+        alias getOverloads = AliasSeq!(__traits(getOverloads, Parent, __traits(identifier, func)));
+    else
+        alias getOverloads = func;
+}
+
+///
+@safe unittest
+{
+    static struct S
+    {
+        int overloadTest(int n = 3) { return n; }
+        int overloadTest(int, int)  { return 3; }
+        int overloadTest2(int, int) { return 3; }
+    }
+
+    void overloadTestLocalFunction() {}
+
+    static assert(getOverloads!overloadTestLocalFunction.length == 1);
+
+    static assert(getOverloads!(S.overloadTest).length == 2);
+    static assert(is(typeof(&S.overloadTest) == typeof(&getOverloads!(S.overloadTest)[0])));
+    static assert(!is(typeof(&S.overloadTest) == typeof(&getOverloads!(S.overloadTest)[1])));
+    alias fn = S.overloadTest;
+    static assert(is(typeof(&fn) == int function(int = 3) pure nothrow @nogc @safe));
+    static assert(is(typeof(&getOverloads!(S.overloadTest)[0]) == int function(int  = 3) pure nothrow @nogc @safe));
+    static assert(is(typeof(&getOverloads!(S.overloadTest)[1]) == int function(int, int) pure nothrow @nogc @safe));
+    static assert(is(typeof(&fn) == typeof(&getOverloads!(S.overloadTest)[0])));
+    static assert(!is(typeof(&fn) == typeof(&getOverloads!(S.overloadTest)[1])));
+    static assert(getOverloads!(S.overloadTest2).length == 1);
+    static assert(getOverloads!(() => 3).length == 1);
+    static assert(getOverloads!(() => 3).length == 1);
+}
+
+/**
+ * Gets the overload for `func` that can be called with `Args` as arguments, if any.
+ *
+ * Params:
+ *    func = the overload set to search.
+ *    Args = the set of argument types to look for, specified as types or as values.
+ * Returns:
+ *    The overload that matches the argument types in `Args`.
+ */
+template overloadFor(alias func, Args...)
```

Comment:
Can you please have a look at PR #6885, which I created, unaware of this PR. It basically does the same thing, but also allows to specify function attributes (which may be the only thing distinguishing overloads).

May be a good idea to integrate that into this PR.

Reply:
I've had a look-see, and my opinion is everything #6885 does can be done in a library without big syntax sacrifices. Function attributes can also be handled (but aren't at the moment), as pointed out above by letting the template take a lambda: `exactOverloadFor!(fun, (ref int) => 1)`

Reply:
#6885 does not propose new syntax (how would it, being a Phobos PR?). There's only a comment  there suggesting an alternative solution using ˋ__traitsˋ.

Anyway, both this PR and #6885 seem to have stalled and I'd like to see some solution for getting an alias to a specific overload.

By the way, I like the idea of taking a function (possibly lambda) as a template parameter, but there should probably be a way to just specify types as well, because parameter attributes might be unknown, making it impossible to create a correct lambda.